docs/security.md

```diff
@@ -0,0 +1,199 @@
+
+1. Between the client code, executing in a browser, and the service, executing
+   on a remote server.
+2. Between the annotated content (which may be an HTML page or a PDF rendered as
+   an HTML page) and the client application. This boundary is marked with
+   asterisks (\*) in Figure 1.
+
+
+## Threat model
+
+We are principally interested in ensuring that untrusted parties cannot gain
+access to data that is intended to be confidential, or tamper with such data
+when it is in transit. Protected data might include:
+
+- user credentials
+- annotation data or metadata which is displayed by the client
+- user profile information
+- group membership records
+- user search history
+
+We must assume that the user has a baseline level of trust in:
+
+1. their browser software (and the platform it runs on)
+2. our client software
+3. the annotation service
+4. any 3rd-party account provider mediating access to the annotation service
+   (e.g. Google, Facebook, etc.)
+
+Any other parties are considered untrusted. Untrusted actors thus include any
+and all of the following:
+
+- the publishers of arbitrary web pages (including annotated content)
+- advertisers or other 3rd-party contributors to arbitrary web pages (including
+  annotated content)
+- other users of the annotation service who have not been explicitly designated
+  as trusted (through group membership, for example)
+- members of the public who don't use the annotation service
+- active attackers
+
+We aim to defend confidential user data against any possibility of unauthorised access.
+
+
+## Potential attack vectors
+
+The mechanisms of directed attack we are aiming to defend against are common to
+many web applications, namely:
+
+- execution of untrusted code in a trusted context (principally by
+  [XSS](https://en.wikipedia.org/wiki/Cross-site_scripting))
+- [clickjacking](https://en.wikipedia.org/wiki/Clickjacking)
+- phishing/imitation attacks
+- eavesdropping of unencrypted network traffic by an untrusted party
+- to a limited extent, [cross-site request
+  forgery](https://en.wikipedia.org/wiki/Cross-site_request_forgery), although
+  this is mostly a concern for the annotation service
+
+
+## Design considerations and defenses
+
+### Same-origin policy protections
+
+The starting point for understanding many of the client-side security mechanisms
+is the web platform's [same-origin
+policy](https://en.wikipedia.org/wiki/Same-origin_policy) (SOP), which ensures
+that any document on origin[^1] "A" has very limited access to the execution
+context or DOM tree of any document on a different origin "B".
+
+[^1]: An origin is the tuple of (scheme, host, port) for a given web document.
+
+![](security-sop.png)
+*Figure 2: Distinct origins for annotated content and client application*
+
+As shown in Figure 2, the bulk of the Hypothesis client application executes
+within an `<iframe>` injected into the annotated content. This `<iframe>` has an
```

Comment:
Something which I know you know but I'll note that came up as a point of confusion/surprise in recent discussions in #frontend is that while the iframe's `contentDocument` and `contentWindow` are either not accessible or only accessible in very limited ways to code in the host page, the `<iframe>` _element_ itself can be manipulated by the host page, including redirecting it to a different URL after it loads.